### PR TITLE
release update and blog post for 2.21.0

### PR DESCRIPTION
**Pull request by ianturton**
Please add any other sponsors that you would like to thank.

### CODE REVIEW

File: _posts/2022-05-24-geoserver-2-21-0-released.md [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+WMS Settings have new capability to filter out any mark factories not being used, and provide an order to prioritise the ones being used.
+
+![MarkFactory precedence](/img/posts/2.21/markfactory.png)<br/>
+
+For more information see [WMS Web Administration](https://docs.geoserver.org/stable/en/user/services/wms/webadmin.html#mark-factory-precedence) (user guide).
+
+* [GEOS-10230](https://osgeo-org.atlassian.net/browse/GEOS-10230) MarkFactory WMS rendering performance optimization
+
+Thanks to Fernando Mino (GeoSolutions Group) for troubleshooting this performance issue, and proposal [GSIP-205](https://github.com/geoserver/geoserver/wiki/GSIP-204) as an optimization.
```
Comment (bradh) [inline _posts/2022-05-24-geoserver-2-21-0-released.md:103]:
```suggestion
Thanks to Fernando Mino (GeoSolutions Group) for troubleshooting this performance issue, and proposal [GSIP-204](https://github.com/geoserver/geoserver/wiki/GSIP-204) as an optimization.
```
(I'm assuming this what was meant)

File: _posts/2022-05-24-geoserver-2-21-0-released.md [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+### Log4J 2 Upgrade
+
+The assessment of [Log4Shell]({% post_url 2021-12-13-logj4-rce-statement %}) vulnerability highlighted that although GeoSever was not affected, our use of the older Log4j 1.2 was a notable risk. This discussion resulted a [small fundraising effort]({% post_url 2022-01-20-log4j-upgrade %}) and proposal to [upgrade to Log4j 2](https://github.com/geoserver/geoserver/wiki/GSIP-167).
+
+The result is a small change to the user interface, listing logging profiles by name (previously the file extension was also listed).
+
+![Log4j update](/img/posts/2.21/logging.png)<br/>
+
+Internally this release replaces changes from Log4j 1.2 logging profiles (using ``properties`` extension) to Log4j 2 logging profiles (using ``xml`` extension):
```
Comment (bradh) [inline _posts/2022-05-24-geoserver-2-21-0-released.md:113]:
```suggestion
Internally this release changes from Log4j 1.2 logging profiles (using ``properties`` extension) to Log4j 2 logging profiles (using ``xml`` extension):
```

File: _posts/2022-05-24-geoserver-2-21-0-released.md [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+  A customization to ``PRODUCTION_LOGGING.properties`` will be backed up to ``PRODUCTION_LOGGING.properties.bak``. This can be restored by renaming ``PRODUCTION_LOGGING.properties.bak`` to ``CUSTOM_LOGGING.properties``.
+
+In addition to the ``INFO`` status messages, you will notice a new ``CONFIG`` logging level used during application startup:
+```
+CONFIG [org.geoserver] - GeoServer configuration lock is enabled
+CONFIG [org.geoserver] - Loading catalog...
+...
+```
+
+For more information, and examples of writing on log4J 2 profile, see [Logging Settings](https://docs.geoserver.org/latest/en/user/configuration/globalsettings.html#logging-settings) and [Advanced log configuration](https://docs.geoserver.org/latest/en/user/configuration/logging.html) in the User Guide. Of note is the introduction of a new ``CONFIG`` logging level used loading and saving configuration changes.
```
Comment (bradh) [inline _posts/2022-05-24-geoserver-2-21-0-released.md:130]:
```suggestion
For more information, and examples of writing a Log4J 2 profile, see [Logging Settings](https://docs.geoserver.org/latest/en/user/configuration/globalsettings.html#logging-settings) and [Advanced log configuration](https://docs.geoserver.org/latest/en/user/configuration/logging.html) in the User Guide. Of note is the introduction of a new ``CONFIG`` logging level used when loading and saving configuration changes.
```

File: _posts/2022-05-24-geoserver-2-21-0-released.md [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+* [GEOS-10426](https://osgeo-org.atlassian.net/browse/GEOS-10426) GISP 167: Upgrade Log4j
```
Comment (bradh) [inline _posts/2022-05-24-geoserver-2-21-0-released.md:132]:
```suggestion
* [GEOS-10426](https://osgeo-org.atlassian.net/browse/GEOS-10426) GSIP 167: Upgrade Log4j
```

File: _posts/2022-05-24-geoserver-2-21-0-released.md [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+  
+  KML can now be used with WPS service for both input and output parameters.
+  
+* [GEOS-10391](https://osgeo-org.atlassian.net/browse/GEOS-10391) Add external output directory setting to limit where processes can write
+  
+  See WPS setting for [external output directory](https://docs.geoserver.org/latest/en/user/services/wps/administration.html#execution-and-resource-management-options) in User Guide.
+  
+* [GEOS-10431](https://osgeo-org.atlassian.net/browse/GEOS-10431) Add WPS setting to disable remote complex inputs.
+  
+  See [WPS Security and input limits](https://docs.geoserver.org/main/en/user/services/wps/security.html#complex-inputs) in User Guide.
```
Comment (bradh) [inline _posts/2022-05-24-geoserver-2-21-0-released.md:163]:
```suggestion
  See [WPS Security and input limits](https://docs.geoserver.org/main/en/user/services/wps/security.html#complex-inputs) in the User Guide.
```

**Review by bradh — COMMENTED** [review]
A few minor comments.

File: _posts/2022-05-24-geoserver-2-21-0-released.md [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+  KML can now be used with WPS service for both input and output parameters.
+  
+* [GEOS-10391](https://osgeo-org.atlassian.net/browse/GEOS-10391) Add external output directory setting to limit where processes can write
+  
+  See WPS setting for [external output directory](https://docs.geoserver.org/latest/en/user/services/wps/administration.html#execution-and-resource-management-options) in User Guide.
+  
+* [GEOS-10431](https://osgeo-org.atlassian.net/browse/GEOS-10431) Add WPS setting to disable remote complex inputs.
+  
+  See [WPS Security and input limits](https://docs.geoserver.org/main/en/user/services/wps/security.html#complex-inputs) in the User Guide.
+
```
Comment (jodygarnett) [inline _posts/2022-05-24-geoserver-2-21-0-released.md:164]:
```suggestion

Thanks to the GeoNode project for these WPS improvements.
```

**Review by jodygarnett — CHANGES_REQUESTED** [review]
Minor suggestions below, really we want to provide an opportunity for @simboss and @aaime to thank any customers interested in being acknowledged.

File: _posts/2022-05-24-geoserver-2-21-0-released.md [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+The above example works around the limitations of shapefile to use longer names, and creates a new attribute ``capital`` on the fly from an expression, as shown in the following GetFeatureInfo output.
+
+![GetFeatureInfo tasmania_cities](/img/posts/2.21/tasmania_feature_type.png) <br/>
+
+This is a great new addition to GeoServer, please see [Feture Type Details](https://docs.geoserver.org/latest/en/user/data/webadmin/layers.html#feature-type-details-vector) in the user guide for details.
+
+* [GEOS-10356](https://osgeo-org.atlassian.net/browse/GEOS-10356) Allow feature type customization
+
+Thanks to Andrea Aime (GeoSolutions) for proposal [GSIP-207](https://github.com/geoserver/geoserver/wiki/GSIP-207) and implementation.
```
Comment (aaime) [inline _posts/2022-05-24-geoserver-2-21-0-released.md:42]:
```suggestion
Thanks to Andrea Aime (GeoSolutions) for proposal [GSIP-207](https://github.com/geoserver/geoserver/wiki/GSIP-207) and implementation, and Canton of Solothurn for funding the work.
```

File: _posts/2022-05-24-geoserver-2-21-0-released.md [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+* [how2map.com](http://how2map.com)
+* [www.geonovation.nl](https://www.geonovation.nl/)
+
+### Logging REST API
+
+For more information please see [Logging settings](https://docs.geoserver.org/latest/en/user/rest/api/logging.html) (User Guide) and [GeoServer Logging](https://docs.geoserver.org/latest/en/api/#1.0.0/logging.yaml) (REST API).
+
+* [GEOS-10368](https://osgeo-org.atlassian.net/browse/GEOS-10368) Logging Controller Addition allows configuration of logging via REST API.
+
+Thanks to Yalın Eren Deliorman for this contribution.
```
Comment (aaime) [inline _posts/2022-05-24-geoserver-2-21-0-released.md:147]:
```suggestion
Thanks to Yalın Eren Deliorman (GeoSolutions) for this contribution, and Eumetsat for sponsoring the work.
```

File: _posts/2022-05-24-geoserver-2-21-0-released.md [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+
+### New WPS settings and KML input/output support
+
+A number of improvements have been made to the WPS service:
+
+* [GEOS-10443](https://osgeo-org.atlassian.net/browse/GEOS-10443) Graduate kml-ppio community module to wps extension
```
Comment (aaime) [inline _posts/2022-05-24-geoserver-2-21-0-released.md:153]:
```suggestion
* [GEOS-10443](https://osgeo-org.atlassian.net/browse/GEOS-10443) Graduate kml-ppio community module to wps extension. Thanks to GeoSolutions for taking the maintainership of the module, and Masego Inc. for sponsoring the graduation work.
```

File: _posts/2022-05-24-geoserver-2-21-0-released.md [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+
+WMS Settings have new capability to filter out any mark factories not being used, and provide an order to prioritise the ones being used.
+
+![MarkFactory precedence](/img/posts/2.21/markfactory.png)<br/>
+
+For more information see [WMS Web Administration](https://docs.geoserver.org/stable/en/user/services/wms/webadmin.html#mark-factory-precedence) (user guide).
+
+* [GEOS-10230](https://osgeo-org.atlassian.net/browse/GEOS-10230) MarkFactory WMS rendering performance optimization
+
+Thanks to Fernando Mino (GeoSolutions Group) for troubleshooting this performance issue, and proposal [GSIP-204](https://github.com/geoserver/geoserver/wiki/GSIP-204) as an optimization.
```
Comment (aaime) [inline _posts/2022-05-24-geoserver-2-21-0-released.md:103]:
```suggestion
Thanks to Fernando Mino (GeoSolutions Group) for troubleshooting this performance issue, and proposal [GSIP-204](https://github.com/geoserver/geoserver/wiki/GSIP-204) as an optimization. Thanks to EMSA for sponsoring the improvement.
```

**Review by aaime — COMMENTED** [review]
Some more sponsors thanks.

**Pull request merged**